bpo-44660: Updated email.feedparser with support for message/global emails with quoted-printable and base64 Content-Transfer-Encodings.

f18a14c09s opened the pull request:
Fixed issue 44660: email.feedparser Module Lacks Support for Section 3.5 of RFC 6532: message/global Emails with non-identity Content-Transfer-Encodings as follows:
Added message/global Content-Transfer-Encoding section to _parsegen function; three supporting classes; and a supporting function.
Added unit test coverage for the subject scenarios.


<!-- issue-number: [bpo-44660](https://bugs.python.org/issue44660) -->
https://bugs.python.org/issue44660
<!-- /issue-number -->

f18a14c09s committed:
Initial version of fix.

Added message/global Content-Transfer-Encoding section to _parsegen function, three supporting classes, and a supporting function.

f18a14c09s committed:
Added parsing unit test coverage.

Added parsing coverage for base64 and quoted-printable Content-Transfer-Encodings.

f18a14c09s committed:
Corrected comment.

jdevries3133 commented on Lib/email/feedparser.py:
```diff
@@ -534,3 +559,85 @@ class BytesFeedParser(FeedParser):
+
+
+class NonIdentityTransferDecodingFeedParser(abc.ABC):
+    """
```
hanging indent on the docstring looks strange to me; consult [PEP 257](https://www.python.org/dev/peps/pep-0257/)

jdevries3133 commented on Lib/email/feedparser.py:
```diff
@@ -534,3 +559,85 @@ class BytesFeedParser(FeedParser):
+
+class Base64TransferDecodingFeedParser(NonIdentityTransferDecodingFeedParser):
+    """
+    Concrete, "base64" implementation of TransferDecodingFeedParser.
```
This docstring, and the docstring for `QuotedPrintableTransferDecodingFeedParser` seem redundant. Seeing as the class and base class names are quite descriptive, maybe just provide an RFC (or other) citation in the docstrings?

f18a14c09s replied:
Thanks, I incorporated the recommendation in a massive update I'm about to apply.

jdevries3133 commented on Lib/email/feedparser.py:
```diff
@@ -534,3 +559,85 @@ class BytesFeedParser(FeedParser):
+    Creates a new FeedParser object that transparently reverses the specified
+     content_transfer_encoding transformation.
+    :param content_transfer_encoding: str
+    :return: Union[NoneType,NonIdentityTransferDecodingFeedParser] None if
```
If you're going to provide this information, use type hints:

```python
def _new_transfer_decoding_parser(content_transfer_encoding: str, **kwargs
    ) -> Union[NoneType, NonIdentityTransferDecodingFeedParser]: 
```
Formatting is tricky because the bass class name is so long. Maybe consider shortening it a bit: `NoIdTransDecodingFP`. I don't really love that abbreviation myself and I like the current descriptive name, but it is quite a mouthful.

f18a14c09s replied:
Thanks, I incorporated the recommendation in a massive update I'm about to apply.

jdevries3133 reviewed (commented):
Very nice!

f18a14c09s committed:
Productionization and attempt to address jdevries3133's recommendations.

Replaced decoding-parser factory method with Content-Transfer-Encoding (str) to factory (function) map.
Removed the word "transfer" from the TransferDecoding-related classes, functions, variables, etc. to be more concise and because there is already sufficient context.
Removed the word "nonidentity" from class NonIdentityTransferDecodingFeedParser's name and updated the docstrings/comments explaining why the word is redundant.
Moved base-64-only functionality to new class in base64mime sub-package.
Accounted for edge cases in QuotedPrintableFeedParser so that it could be made a standalone feature if desired.
Added/updated code comments.

f18a14c09s committed:
Added and refactored the error handling.

f18a14c09s committed:
Added/updated the comments/docstrings.

f18a14c09s committed:
Removed unused, commented-out code.

f18a14c09s committed:
Merge pull request #1 from f18a14c09s/issue-44660-code-review-changes-1

Issue 44660 code review changes 1
